### DISCUSSION TITLE
Compiling groops with custon LAPACK installation

### DISCUSSION THREAD

**Discussion by AndreHauschild** [groops-devs/groops#260] [Show and tell]
Dear groops community,

for testing purposes I had to compile groops with a LAPACK version other than the system version. I would like to share the necessary modifications to `CMakeList.txt` here, because it might be of interest for other users as well.

Assuming that you have downloaded the LAPACK source files and compiled them in `$(HOME)/Software/`, you must set the path to the libraries like so

```
set(HOME_DIR $ENV{HOME})
set(LAPACK_DIR "${HOME_DIR}/Software/lapack-3.12.0")
set(BLAS_LIBRARIES "${LAPACK_DIR}/librefblas.a")
set(LAPACK_LIBRARIES "${LAPACK_DIR}/liblapack.a;-lm;-ldl")
```

and comment out the search for the system libraries

```
#find_package(BLAS   REQUIRED)
#find_package(LAPACK REQUIRED)
```

Then, also change the order of the library inclusion to avoid linker errors

```
set(BASE_LIBRARIES ${LAPACK_LIBRARIES} ${BLAS_LIBRARIES} ${EXPAT_LIBRARIES} stdc++fs)
```

Here is a snippet of the relevant part of  `CMakeList.txt` with all the changes:

```
# External Source Files
# ---------------------
# HWM14     Horizontal Wind Model 2014 (https://map.nrl.navy.mil/map/pub/nrl/HWM/HWM14/)
# NRLMSIS   NRLMSIS 2.0 thermospheric model (https://map.nrl.navy.mil/map/pub/nrl/NRLMSIS/NRLMSIS2.0/)
# JB2008    JB2008 thermospheric model (http://sol.spacenvironment.net/jb2008/)
# IGRF      International Geomagnetic Reference Field (https://doi.org/10.1186/s40623-015-0228-9)
# IERS      International Earth Rotation and Reference Systems Service (IERS) Conventions software collection (https://iers-conventions.obspm.fr/)

# Specify the full paths to the custom BLAS and LAPACK libraries
set(HOME_DIR $ENV{HOME})
set(LAPACK_DIR "${HOME_DIR}/Software/lapack-3.12.0")
set(BLAS_LIBRARIES "${LAPACK_DIR}/librefblas.a")
set(LAPACK_LIBRARIES "${LAPACK_DIR}/liblapack.a;-lm;-ldl")

#find_package(BLAS   REQUIRED)
#find_package(LAPACK REQUIRED)
find_package(EXPAT  REQUIRED)
include_directories(${EXPAT_INCLUDE_DIRS})

set(BASE_LIBRARIES ${LAPACK_LIBRARIES} ${BLAS_LIBRARIES} ${EXPAT_LIBRARIES} stdc++fs)
```

Hope this helps!

**Comment by Hanbing59** [comment]
Hi Andre,

thanks for your sharing.

From you post

> also change the order of the library inclusion to avoid linker errors

just out of curiosity, why should we use a different order for the library inclusion when using custom `LAPACK/BLAS` library?

**Reply by AndreHauschild** [reply]
Hi Hanbing, 

admittedly I do not know the reason, but it turned out to be necessary to avoid linker errors.

**Reply by Hanbing59** [reply]
Hmm, interesting.
